gh-126742: allow to use non-UTF8 exception messages

picnixz opened the pull request:
<!-- gh-issue-number: gh-126742 -->
* Issue: gh-126742
<!-- /gh-issue-number -->

picnixz committed:
allow to use translated exception messages

ZeroIntensity commented on Python/errors.c:
```diff
@@ -301,6 +301,23 @@ PyErr_SetString(PyObject *exception, const char *string)
+_PyErr_SetLocaleStringTstate(PyThreadState *tstate, PyObject *exception,
+                             const char *string)
+{
+    PyObject *value = PyUnicode_DecodeLocale(string, "strict");
```
I think this should be `surrogateescape`, not `strict`:
```suggestion
    PyObject *value = PyUnicode_DecodeLocale(string, "surrogateescape");
```

picnixz replied:
Both are acceptable and I think it's preferrable to use a strict decoding since a message that is not acceptable by the current locale feels wrong. In addition, using surrogateescape means that for re-encoding the message you need to also pass the "surrogateescape" handler.

picnixz replied:
Ok, looking around, other usages in the context of error rendering use the "surrogateescape" handler, so I've decided to use it as well.

ZeroIntensity commented on Python/errors.c:
```diff
@@ -301,6 +301,23 @@ PyErr_SetString(PyObject *exception, const char *string)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    _PyErr_SetLocaleStringTstate(tstate, exception, string);
+}
```
It might be worth just turning this into an inline function or macro.

picnixz replied:
Others functions are implemented without inlining and without macros so I think it's fine to keep it that way here.

picnixz committed:
Use 'surrogateescape' handler instead of 'strict' handler.

vstinner commented on Include/internal/pycore_pyerrors.h:
```diff
@@ -130,6 +130,25 @@ PyAPI_FUNC(void) _PyErr_SetString(
     const char *string);
 
+/*
+ * Similar to _PyErr_SetString but decodes 'string' into the current locale.
```
```suggestion
 * Similar to _PyErr_SetString but decodes 'string' from the current locale.
```

vstinner reviewed (commented):
I don't see the purpose of these internal functions since they are now used. This PR only adds dead code. I would prefer to see how these functions are used.

If only dlerror() uses it, why not using code using dlerror()?

picnixz committed:
fix typos

picnixz committed:
Merge branch 'main' into feat/capi/set-locale-exception-126742

picnixz committed:
Allow to set localized error string.

picnixz committed:
Use `PyErr_SetLocaleString` when possible

picnixz committed:
Indicate which external functions may return non-UTF8 error strings.

picnixz committed:
Merge branch 'main' into feat/capi/set-locale-exception-126742

picnixz committed:
fix build

picnixz committed:
update dll

picnixz committed:
add docs & news

picnixz committed:
update comment

picnixz committed:
update TODO

picnixz committed:
update comment

picnixz committed:
Revert "add docs & news"

This reverts commit 16a98a290b2b633b193cb487345e1215b9ac3bb8.

picnixz committed:
keep one internal helper instead of a public/private API pair

picnixz committed:
rename references to `PyErr_SetLocaleString`

picnixz committed:
Simplify logic

picnixz committed:
Merge branch 'main' into fix/locale-set-object-exception-126742

vstinner commented on Doc/data/refcounts.dat:
```diff
@@ -724,6 +724,10 @@ PyErr_SetImportErrorSubclass:PyObject*:path:+1:
 
+_PyErr_SetLocaleString:int:::
+_PyErr_SetLocaleString:PyObject*:type:+1:
+_PyErr_SetLocaleString:const char*:message::
```
I'm not sure that it's useful to document the refcount behavior of internal functions. Other functions prefixed by _Py are part of the public C API and/or documented.

picnixz replied:
I don't think it hurts (or we can have a separate file for that). My motivation is to move towards a maintained `refcounts.dat` for every function, even those that are not part of the Stable ABI, but if you think it's better to do it later, I can remove that entry.

picnixz replied:
I've decided to remove the entry so that we can add it in one go later.

vstinner commented on Include/internal/pycore_pyerrors.h:
```diff
@@ -130,6 +130,17 @@ PyAPI_FUNC(void) _PyErr_SetString(
+ * in which case, this returns -1. Otherwise this returns 0 if the localized
+ * exception has been successfully set.
+ *
+ * Exported for shared extensions but not part of the stable ABI.
```
```suggestion
 * Exported for '_ctypes' shared extensions.
```

picnixz replied:
Oh, so "_ctypes" shared extensions means other shared extensions?

vstinner reviewed (commented):
Is there a practical way to get a non-ASCII error messages not encoded to UTF-8 for the following functions?

* dlerror()
* gdbm_strerror()

How can I test (manually) modified code paths?

picnixz committed:
Remove entry from refcounts.dat.

We will add internal functions later in one go instead one-by-one.

picnixz committed:
update comment

vstinner commented on Include/internal/pycore_pyerrors.h:
```diff
@@ -130,6 +130,17 @@ PyAPI_FUNC(void) _PyErr_SetString(
     const char *string);
 
+/*
+ * Set an exception with the error message decoded from the current locale.
```
```suggestion
 * Set an exception with the error message decoded from the current locale
 * encoding (LC_CTYPE).
```

vstinner reviewed (approved):
The change LGTM, but I didn't test it (manually).

serhiy-storchaka reviewed (commented):
Translated error messages most likely depend on the LC_MESSAGE environment variable. It is a separate difficult issue if its encoding differs from LC_CTYPE's encoding.

But if it is non-ASCII because it includes file paths, we should use the FS encoding.

picnixz committed:
Update Include/internal/pycore_pyerrors.h

Co-authored-by: Victor Stinner <vstinner@python.org>

picnixz commented on Include/internal/pycore_pyerrors.h:
```diff
@@ -130,6 +130,18 @@ PyAPI_FUNC(void) _PyErr_SetString(
+ *
+ * Exceptions occurring in decoding take priority over the desired exception,
+ * in which case, this returns -1. Otherwise this returns 0 if the localized
+ * exception has been successfully set.
```
```suggestion
 * Exceptions occurring in decoding take priority over the desired exception.
 * In those cases, this function returns -1. Otherwise this returns 0 if the
 * localized exception has been successfully set.
```

picnixz committed:
fix grammar

picnixz committed:
Merge branch 'main' into fix/locale-set-object-exception-126742

picnixz committed:
Unconditionally re-raise decoding errors

serhiy-storchaka reviewed (commented):
Could you please add tests?

picnixz committed:
add tests for `_PyErr_SetLocaleString`

picnixz committed:
add tests for `dlerror` and `gdbm_*` functions

vstinner commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +123,17 @@ def test_null_dlsym(self):
+    def test_localized_error(self):
+        with self.assertRaisesRegex(
+            OSError,
+            re.escape("foo.so: Ne peut ouvrir le fichier d'objet partagé"),
```
@serhiy-storchaka: We don't control dlerror() error message and translated error message. The test fails if translations are not installed or if the error message changes.

I'm not sure if it's a good idea to rely on the exact dlopen() translated error message.

serhiy-storchaka replied:
We should not rely on the dlopen() error message. We only need to check that the error is not a UnicodeDecodeError.

Try to use a UTF8-undecodable bytes name, e.g. `b'\xff'` (bytes are accepted, right?).

picnixz replied:
Yes, and since I don't need to create it, I don't encounter issues with such filename. Depending on whether the locale is ISO-88591 or not, `\xff` decodes differently so it's fine (note that the OSError being raised does not have the error message as bytes but a string, so I can't check against bytes directly (I think?)).

serhiy-storchaka commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +123,17 @@ def test_null_dlsym(self):
             self.assertEqual(os.read(pipe_r, 2), b'OK')
 
 
+class TestCAPI(unittest.TestCase):
```
This is not a C API test.

picnixz replied:
I've renamed it to `TestLocalization` (or something like that).

serhiy-storchaka commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +123,17 @@ def test_null_dlsym(self):
+class TestCAPI(unittest.TestCase):
+
+    @unittest.skipUnless(hasattr(_ctypes, 'dlopen'), 'require ctypes.dlopen()')
+    @test.support.run_with_locales('LC_ALL', 'fr_FR.utf8', 'fr_FR.iso88591')
```
I think that `run_with_locale()` instead of `run_with_locales()` can be enough. But try to use non-UTF8 locales first. Try to use several locales, not only French. Add `''` at the end to use the current locale as a fallback.

picnixz replied:
I've chosen fr_FR.iso88591, ja_JP.sjis and zh_CN.gbk. Do you think others should be used?

serhiy-storchaka commented on Lib/test/test_dbm_gnu.py:
```diff
@@ -205,6 +207,11 @@ def test_clear(self):
 
+    @support.run_with_locales('LC_ALL', 'fr_FR.UTF-8', 'fr_FR.iso88591')
+    def test_localized_error(self):
+        expect = re.escape('Base de données vide')
```
We should not rely on the error message. We only need to check that the error is not a UnicodeDecodeError.

serhiy-storchaka commented on Lib/test/test_dbm_gnu.py:
```diff
@@ -205,6 +207,11 @@ def test_clear(self):
+    def test_localized_error(self):
+        expect = re.escape('Base de données vide')
+        empty = create_empty_file(filename)
+        self.assertRaisesRegex(gdbm.error, expect, gdbm.open, filename, 'r')
```
filename is TESTFN. It can be not encodable with the tested locale (we should try non-UTF8 locales first). Try to use ASCII-only name like "nonexisting". Try to use a UTF8-undecodable bytes name if bytes is accepted, otherwise use a name containing a surrogateescaped byte.

picnixz replied:
Unfortunately, non-existing ASCII name wouldn't work because the error would be an errno error instead of a `gdbm.error`. I really need a specific kind of error to trigger in order to hit the translated messages. But I can't create an empty file with UTF8-undecodable byte names I think (os.open() fails with a ValueError). So I'm not entirely sure how I can do it =/ I'll try to find other ways offline (I will be offline until Friday)

serhiy-storchaka replied:
`open(b'\xff')`

picnixz replied:
How come `create_empty_file` failed when I tried? I'll need to check it later. But thanks! (I remember trying using the FN_NONDECODABLE file but didn't try with just '\xff').

vstinner replied:
Why not testing an existing empty file with an ASCII filename? Something like:

```py
filename = "empty"
open(filename, "wb").close()
self.addCleanup(os_helper.unlink, filename)
# ... use filename ...
```

serhiy-storchaka reviewed (commented):
I do not think that we need a C API test. It is just a private helper. In any case, the C API test is not working.

picnixz committed:
fix tests

picnixz committed:
cosmetic changes on imports

picnixz committed:
remove C API internal tests

picnixz committed:
fix tests

serhiy-storchaka commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +122,28 @@ def test_null_dlsym(self):
 
 
+@unittest.skipUnless(sys.platform.startswith('linux'),
+                     'test requires _ctypes.dlopen()')
```
Why does it require Linux? It only should require `_ctypes.dlopen`.

serhiy-storchaka commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +122,28 @@ def test_null_dlsym(self):
+
+    @test.support.run_with_locale(
+        'LC_ALL',
+        'fr_FR.iso88591', 'ja_JP.sjis', 'zh_CN.gbk',
```
Add also UTF-8 locales.

serhiy-storchaka commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +122,28 @@ def test_null_dlsym(self):
+        # encode '\xff' as '\udcff' or 'ÿ', but we are only interested in
+        # avoiding a UnicodeDecodeError when reporting the dlerror() error
+        # message which contains the localized filename.
+        filename_pattern = r'missing[ÿ|\udcff].so:.+'
```
This is not always correct for other 8-bit locales. `b'\xff'` is not always decoded to `'\xff'`.

serhiy-storchaka commented on Lib/test/test_dbm_gnu.py:
```diff
@@ -205,6 +206,15 @@ def test_clear(self):
+    )
+    def test_localized_error(self):
+        with temp_dir() as d:
+            create_empty_file(os.path.join(d, TESTFN))
```
TESTFN is non-ASCII which is encodable with the filesystem encoding when tests are imported. It may be not encodable in the specified test locale. So I suggest to use name containing `'\udcff'` or something like.

It may be impossible to create a non-UTF-8 file name on macOS, so catch an error in `create_empty_file()` and skip the test.

You may need to test this on Windows manually to ensure that the test works and not always skipped (if gdbm is available on Windows).

picnixz replied:
Ah I thought it would have been file. My question: since I don't care about the filename in this case (the error doesn't report it), I can just use an ASCII name and that's it right? (I'm just interested in not having UnicodeDecodeError)

serhiy-storchaka replied:
Yes, you can just use an ASCII name.

serhiy-storchaka commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -984,15 +984,8 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
-        // Ignore errors from PyUnicode_DecodeLocale,
-        // fall back to the generic error below.
-        PyErr_Clear();
+        _PyErr_SetLocaleString(PyExc_ValueError, dlerr);
```
Can this be tested too?

serhiy-storchaka commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -3805,21 +3798,14 @@ PyCFuncPtr_FromDll(PyTypeObject *type, PyObject *args, PyObject *kwds)
-            // Ignore errors from PyUnicode_DecodeLocale,
-            // fall back to the generic error below.
-            PyErr_Clear();
+            _PyErr_SetLocaleString(PyExc_AttributeError, dlerr);
```
Can this be tested too?

picnixz replied:
I've added a test. Actually it was implicitly tested in `TestNullDlsym`, but only with the current locale.

serhiy-storchaka commented on Modules/main.c:
```diff
@@ -374,6 +374,7 @@ pymain_run_file_obj(PyObject *program_name, PyObject *filename,
     if (fp == NULL) {
         // Ignore the OSError
         PyErr_Clear();
+        // TODO(picnixz): strerror() is locale dependent but not PySys_FormatStderr().
```
Then use `PyUnicode_DecodeLocale()` and add a test.

Other idea: use `str()` of the current error (before clearing it). It is not necessary `OSError`, so `errno` and `strerror()` may be wrong.

I do not think there is a simple way to test this, besides using embedded Python. So it is okay to not add a test for this.

picnixz replied:
Do you want me to actually address this in this PR or in a follow-up PR?

serhiy-storchaka replied:
It is up to you.

picnixz committed:
fix `dbm` tests

picnixz committed:
improve test coverage

picnixz committed:
fix macOS tests

picnixz committed:
update names

picnixz committed:
fix tests

picnixz committed:
fix tests (again???)

picnixz committed:
fix tests (macOS)

picnixz committed:
fix tests (windows)

picnixz committed:
fix macOS tests

picnixz committed:
fix windows tests

picnixz committed:
revert a missing line

picnixz committed:
revert changes

serhiy-storchaka commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -111,13 +110,90 @@ def test_null_dlsym(self):
+    @configure_locales
+    def test_localized_error_from_dll(self):
+        with temp_dir() as outdir:
+            dstname = self.make_empty_lib(outdir, 'test_from_dll.so')
```
Could not we use an existing library for these tests? They should not require gcc.

See for example test_loading.

picnixz replied:
Yes, we can do it.

picnixz committed:
use libc.so instead

vstinner commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -111,13 +108,74 @@ def test_null_dlsym(self):
+    def test_localized_error_from_dll(self):
+        dll = CDLL(self.libc_file)
+        with self.assertRaises(AttributeError) as cm:
+            dll.foo
```
```suggestion
            dll.this_name_doesnt_exist
```

vstinner commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -111,13 +108,74 @@ def test_null_dlsym(self):
+    def test_localized_error_in_dll(self):
+        dll = CDLL(self.libc_file)
+        with self.assertRaises(ValueError) as cm:
+            c_int.in_dll(dll, 'foo')
```
```suggestion
            c_int.in_dll(dll, 'this_name_doesnt_exist')
```

vstinner commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -111,13 +108,74 @@ def test_null_dlsym(self):
+    @unittest.skipUnless(hasattr(_ctypes, 'dlopen'),
+                         'test requires _ctypes.dlopen()')
+    @unittest.skipUnless(hasattr(_ctypes, 'dlsym'),
+                         'test requires _ctypes.dlsym()')
```
Does test_null_dlsym() need the same decorators? PyPy doesn't seem to have _ctypes.dlsym() for example.

picnixz replied:
Err... I don't know. For safety, I'll add those decorators as well.

picnixz replied:
FTR, I decided to only skip the calls that use `dlopen` and `dlsym` but keep the calls that use `CDLL` and `c_int.in_dll` instead of skipping the entire test.

vstinner commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -111,13 +108,74 @@ def test_null_dlsym(self):
+    def test_localized_error_dlsym(self):
+        dll = _ctypes.dlopen(self.libc_file)
+        with self.assertRaises(OSError) as cm:
+            _ctypes.dlsym(dll, 'foo')
```
```suggestion
            _ctypes.dlsym(dll, 'this_name_doesnt_exist')
```

vstinner commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -111,13 +108,74 @@ def test_null_dlsym(self):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.libc_file = find_library("c")
```
You may rename if to `libc_filename`.

picnixz committed:
address Victor's review

- replace possible unknown symbol by a better unknown symbol (hopefully)
- improve semantic naming ('libc_file' -> 'libc_filename')
- ensure PyPy compatibility

vstinner reviewed (approved):
LGTM

encukou commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +118,66 @@ def test_null_dlsym(self):
+
+    @staticmethod
+    def configure_locales(func):
+        return test.support.run_with_locale(
```
Do you want `run_with_locales`, the decorator?
`run_with_locale` is a context manager.

picnixz replied:
OH! we should fix this then: https://docs.python.org/3/library/test.html#test.support.run_with_locale. And I wanted the decorator. Hopefully it will work...

encukou replied:
Oh, never mind, it's both. Sorry for the noise.

picnixz replied:
I've actually changed the test so that all locales are tested and not only the first locale that could be found. Now, the tests are much faster to run so it's probably fine.

encukou replied:
Hm, I'm not sure what the intended difference is. But, `run_with_locales` uses `subTest`, so at least it gives better error reporting.

picnixz replied:
- `run_with_locale` only uses the first locale that is correct
- `run_with_locales` runs for all specified locales (and skip those that cannot be found)
-

serhiy-storchaka replied:
It is both a decorator and a context manager.

The difference is that `run_with_locales` runs the test with all locales and can only be used as a decorator, `run_with_locale` only runs the test with the first available locale and can also be used as a context manager.

In this case it is enough to use `run_with_locale`.

picnixz replied:
Oh.. so you still want to use `run_with_locale`. Mmh, ok.

serhiy-storchaka replied:
Most tests use `run_with_locale`. I added `run_with_locales` recently specially for tests which should be run for several locales.

Perhaps in these cases we should use a different approach and continue to run the test for different locales until we find one with translations, and mark the test skipped if no such locale is found. I'll think about this idea if I have a chance. For now, you can use what you prefer.

picnixz replied:
(I'll keep `run_with_locale` since I don't want yet another edit and since it's probably fine to only test a single locale for those specific tests)

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +118,66 @@ def test_null_dlsym(self):
+
+    @staticmethod
+    def configure_locales(func):
+        return test.support.run_with_locale(
```
```suggestion
        return test.support.run_with_locales(
```

picnixz commented on Lib/test/test_dbm_gnu.py:
```diff
@@ -205,6 +206,16 @@ def test_clear(self):
                 self.assertNotIn(k, db)
             self.assertEqual(len(db), 0)
 
+    @support.run_with_locale(
```
```suggestion
    @support.run_with_locales(
```

picnixz committed:
Fix tests (hopefully it will still work)

picnixz commented on Lib/test/test_ctypes/test_dlerror.py:
```diff
@@ -119,5 +118,66 @@ def test_null_dlsym(self):
+
+    @staticmethod
+    def configure_locales(func):
+        return test.support.run_with_locales(
```
```suggestion
        return test.support.run_with_locale(
```

picnixz commented on Lib/test/test_dbm_gnu.py:
```diff
@@ -205,6 +206,16 @@ def test_clear(self):
                 self.assertNotIn(k, db)
             self.assertEqual(len(db), 0)
 
+    @support.run_with_locales(
```
```suggestion
    @support.run_with_locale(
```

picnixz committed:
Reverting last commit to use `run_with_locale` only

`run_with_locales` is a decorator-only but would run all specified locales,
skipping those that cannot be found. Instead, we use `run_with_locale`
which is a context manager *and* a decorator to speed-up tests.
